src/main/java/com/google/jenkins/plugins/storage/util/RetryStorageOperation.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jenkins.plugins.storage.util;
+
+import static java.util.logging.Level.SEVERE;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.logging.Logger;
+
+import static com.google.api.client.http.HttpStatusCodes
+    .STATUS_CODE_UNAUTHORIZED;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.http.HttpResponseException;
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+import com.google.jenkins.plugins.storage.Messages;
+import com.google.jenkins.plugins.storage.UploadException;
+import com.google.jenkins.plugins.util.Executor;
+import com.google.jenkins.plugins.util.ExecutorException;
+
+import hudson.FilePath;
+import hudson.Util;
+import hudson.model.AbstractBuild;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+
+/**
+ * A class to fascilitate retries on storage operations.
+ */
+public class RetryStorageOperation {
+  private static final Logger logger =
+      Logger.getLogger(RetryStorageOperation.class.getName());
+
+  /**
+   * An operation to be retried
+   */
+  public interface Action {
```

Comment:
The name "Action" is used in the Jenkins object model, which makes this usage a bit confusing. Maybe call these "Operation" and "RepeatableOperation" instead?

Reply:
Good point! Thanks for the suggestion!